src/test/java/org/hisp/dhis/rules/ProgramRuleVariableTest.java

```diff
@@ -0,0 +1,316 @@
+    {
+        org.hisp.dhis.rules.models.Rule rule = getRule( "V{program_name}" );
+
+        List<RuleEffect> ruleEffects = callEnrollmentRuleEngine( rule );
+
+        assertProgramRuleVariableAssignment( ruleEffects, rule,  PROGRAM_NAME);
+    }
+
+    @Test
+    public void testProgramStageIdProgramVariableIsAssigned() throws Exception
+    {
+        org.hisp.dhis.rules.models.Rule rule = getRule( "V{program_stage_id}" );
+
+        List<RuleEffect> ruleEffects = callEventRuleEngine( rule );
+
+        assertProgramRuleVariableAssignment( ruleEffects, rule,  PROGRAM_STAGE);
+    }
+
+    @Test
+    public void testProgramStageNameProgramVariableIsAssigned() throws Exception
+    {
+        org.hisp.dhis.rules.models.Rule rule = getRule( "V{program_stage_name}" );
+
+        List<RuleEffect> ruleEffects = callEventRuleEngine( rule );
+
+        assertProgramRuleVariableAssignment( ruleEffects, rule,  PROGRAM_STAGE_NAME);
+    }
+
+    @Test
+    public void testTEICountProgramVariableIsAssigned() throws Exception
+    {
+        org.hisp.dhis.rules.models.Rule rule = getRule( "V{tei_count}" );
+
+        List<RuleEffect> ruleEffects = callEnrollmentRuleEngine( rule );
+
+        assertProgramRuleVariableAssignment( ruleEffects, rule,  "1");
+    }
+
+    private org.hisp.dhis.rules.models.Rule getRule(String variable) {
+        RuleAction assignAction = RuleActionAssign.create(null, variable, "${test_data_element}");
+        return org.hisp.dhis.rules.models.Rule.create(null, 1, "true", Arrays.asList(assignAction), "test_program_rule1");
+    }
+
+    private void assertProgramRuleVariableAssignment(List<RuleEffect> ruleEffects, Rule rule, String variableValue) {
+        assertThat( ruleEffects.size() ).isEqualTo(1);
+        assertThat( ruleEffects.get(0).data() ).isEqualTo(variableValue);
+        assertThat( ruleEffects.get(0).ruleAction() ).isEqualTo(rule.actions().get( 0 ));
+    }
+
+    private List<RuleEffect> callEnrollmentRuleEngine( Rule rule )
+        throws Exception
+    {
+        RuleEngine.Builder ruleEngineBuilder = getRuleEngine( Arrays.asList(rule));
+
+        RuleEngine ruleEngine = ruleEngineBuilder.build();
+        return ruleEngine.evaluate( getEnrollment() ).call();
+    }
+
+    private List<RuleEffect> callEventRuleEngine( Rule rule )
+        throws Exception
+    {
+        RuleEngine.Builder ruleEngineBuilder = getRuleEngine( Arrays.asList(rule));
+
+        RuleEvent event = RuleEvent.builder()
+            .event(EVENT_ID)
+            .programStage(PROGRAM_STAGE)
+            .programStageName(PROGRAM_STAGE_NAME)
+            .status(RULE_EVENT_STATUS)
+            .eventDate(EVENT_DATE)
+            .dueDate(DUE_DATE)
+            .organisationUnit(ORGANISATION_UNIT)
+            .organisationUnitCode(ORGANISATION_UNIT_CODE)
+            .dataValues( new ArrayList<RuleDataValue>() )
+            .build();
+
+        RuleEngine ruleEngine = ruleEngineBuilder.enrollment( getEnrollment() ).build();
```

Comment:
In this case, the enrollment should contain one event, we should probably call getEnrollment() and then add the event from above to the enrollment before setting the ruleEngineBuilder context?

Reply:
@Bekkalizer event or enrollment which is part of context cannot be evaluated. Any event or enrollment which is subject to evaluation will have to be passed explicitly to the evaluation function.